### ISSUE TITLE
Support Hotel Wifi

### ISSUE THREAD

**Issue by frioux** [tailscale/tailscale#1634]
**Describe the bug**
I went to a hotel that uses open wifi, DNS interception, and an http login, and instead of `http://example.com` getting intercepted, it just never loaded.  Disabling Tailscale allowed me to log in.

**To Reproduce**
Not sure I can help with this.  If there is some tracing I can do at this hotel I can for the next ~hour, but it seemed pretty typical.

**Expected behavior**
Best case scenario tailscale could detect this situation and allow it.  Alternately maybe it could somehow let me know what's happening?  I'm not sure.

**Version information:**
 - Device: Laptop
 - OS: Linux
 - OS version: Ubuntu 20.04
 - Tailscale version: 1.6.0

**Additional context**
My phone runs tailscale and *didn't* have this problem.  It's an Android.

For fun: I'm at the Brisas Del Mar, Inn at the Beach in Santa Barbara.

**Comment by bradfitz** [comment]
What's the network do without Tailscale? Enter your room number and last name and then the wifi stops being intercepted?

I'd guess we don't get a network change event in that case. Maybe we need to fall back to some polling to wait for the network to stop being a lie.

/cc @danderson @DentonGentry

**Comment by frioux** [comment]
well example.com goes to some non-example.com url and asks for a password (which is on the back of the hotel card.)  I didn't look at the actual mechanism but my *guess* is that the DNS that comes from DHCP gave me a DNS server that does.  FWIW with tailscale I never got the initial MITM'd page, so it seems like TS didn't even attempt that DNS server.

**Comment by danderson** [comment]
Yeah, captive portals typically redirect your HTTP traffic with lying DNS responses, to present a captive portal. Once you've signed in, it notes that your IP/MAC address is allowed and stops doing that.

To properly fix, we'd need to have tailscale do captive portal detection in the background, and do... something if that says we've captured. Not sure what though.

In the meantime, split DNS support in 1.8 will at least help with this, in the sense that you can have MagicDNS without hijacking all DNS on the system, so the captive portal would still work.

**Comment by frioux** [comment]
I agree re split DNS and suspect it will almost completely resolve the issue.  If you care, here's a blog post documenting how some things detect this situation (pretty non-standard) https://www.intraway.com/blog/how-browser-identify-captive-portals

**Comment by sailorfrag** [comment]
I believe captive portals don't usually lie about DNS anymore because then the bad DNS response can get cached. Instead, port 80 to any destination is usually captured by a transparent proxy that is either always in place (and it determines if the client has authenticated each time) or the firewall rules update to bypass the proxy after authentication. DNS to anything other than the captive portal's DNS is usually blocked, however.

Most OSes have captive portal detection because the UX is terrible otherwise, especially since transparent proxies don't interact well with TLS. Typically that's implemented by a HTTP (not HTTPS) request to something that generates a 204 response in the happy case and a captive portal will send something else back (200 or a redirect).

So probing for captive portal with the DHCP-provided DNS server is probably a solid approach.

**Comment by DentonGentry** [comment]
Related: https://github.com/tailscale/tailscale/issues/722

**Comment by DentonGentry** [comment]
Split DNS was delivered in the Tailscale 1.8 release several months ago.

**Comment by felicianotech** [comment]
I can confirm that this also happens with the macOS client and Starbucks WiFi. I'm running Tailscale v1.12.3. Disabling Tailscale, connecting to Starbucks, and then re-enabling Tailscale works fine. I can't really test anything else at the moment since I'm now "approved" for the WiFi but I will in the future.

**Comment by benesch** [comment]
United in-flight WiFi portal is also problematic. I don't pretend to know enough about networking to understand this, but here's what I just observed. I connected to the "United_Wi-Fi" network. Connection is successful. Attempt to load "example.com" redirects to https://www.unitedwifi.com:

```
$ curl example.com -v
*   Trying 93.184.216.34:80...
* Connected to example.com (93.184.216.34) port 80 (#0)
> GET / HTTP/1.1
> Host: example.com
> User-Agent: curl/7.77.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 307 Temporary Redirect
< Location: https://www.unitedwifi.com
< Connection: close
<
* Closing connection 0
```

But following that redirect somehow yields the _external_ https://www.unitedwifi.com page; i.e., what you'd see if you access that domain from the public internet rather than from on a United flight. Tailscale is clearly doing some magic to punch that traffic through to the internet. It took me an embarrassingly long time of searching for the nonexistent "buy WiFi" button before I realized what had happened and disabled Tailscale, at which point I was able to access the in-flight portal and buy WiFi.

**Comment by DentonGentry** [comment]
https://github.com/tailscale/tailscale/issues/1634 and https://github.com/tailscale/tailscale/issues/1377 are somewhat similar in that they need a way for MagicDNS to detect that it is in an environment where the set of upstream DNS servers it has been configured to use cannot possibly work. We might consider solving this similar to what browsers do (like with Chrome's redirect204):
- periodically, when there is some other DNS lookup to be done, also lookup a name where we absolutely know what the answer is supposed to be
- if the answer comes back different, use that to figure out if we're in a captive portal environment or DNS64 environment

**Comment by andrew-d** [comment]
On the Hilton Hotels captive portal, `dig ipv4only.arpa` does not appear to be hijacked:
```
$ dig +short ipv4only.arpa
192.0.0.170
192.0.0.171
```

The same values are returned on my regular home network.

**Comment by andrew-d** [comment]
@alcarithemad pointed out that there's a RFC for captive portal identification in DHCP and IPv6 RAs:
https://datatracker.ietf.org/doc/rfc8910/

**Comment by DentonGentry** [comment]
Since the last update to this issue:
- a `/generate_204` endpoint has been added to the DERP servers.
- `tailscale netcheck` uses this endpoint to check for captive portals

The next steps will be automating more of the handling when a captive portal is detected.

**Comment by pshotton** [comment]
I'm hitting this and can find no workaround.
Lyria train hotspot.
With tailscale off, I can connect to the portal, log in and have access to the internet.
With tailscale on the portal page never opens.
If I connect with tailscale off, and then turn tailscale on, internet access is blocked but I can access my tailscale servers. Turn off tailscale and internet access returns.
Tailscale 1.38.3
Chromeos Version 114.0.5735.119 64 bit
$ tailscale netcheck (with tailscale enabled)
2023/06/26 13:53:49 No DERP map from tailscaled; using default.
fetch prodDERPMap failed: Get "https://controlplane.tailscale.com/derpmap/default": dial tcp: lookup controlplane.tailscale.com on 100.115.92.193:53: read udp 100.115.92.201:53725->100.115.92.193:53: i/o timeout

$ tailscale netcheck (with tailscale disabled)
2023/06/26 13:54:48 No DERP map from tailscaled; using default.

Report:
        * UDP: true
        * IPv4: yes, 109.190.253.13:58796
        * IPv6: no, but OS has support
        * MappingVariesByDestIP: false
        * HairPinning: false
        * PortMapping: 
        * Nearest DERP: Frankfurt
        * DERP latency:
                - fra: 67.7ms  (Frankfurt)
                - etc

This is a real problem.

**Comment by mich2k** [comment]
This issue is still happening today, moreover from linux is quite annoying having to down/up with cli

**Comment by sailorfrag** [comment]
When problems occur, what sort of configuration is enabled?

It is much more difficult to design a way for captive portals to work when you have an exit node configured than without, for example. I don't expect captive portals should cause problems if "override local DNS" is disabled (which is the default) and no exit node is selected.

**Comment by ronaldheft** [comment]
I put in the following split DNS config for WiFi on United Airlines.

![image](https://github.com/tailscale/tailscale/assets/1143112/4c60ed7b-1198-4a47-af12-656c309283fc)

**Comment by hakusaro** [comment]
I'm not sure that it's helpful, but I will add a tiny bit of context from my experience using JR East Free WiFi with Tailscale on macOS 15.2 / Tailscale 1.80.1 using MagicDNS. Normally, obviously, I know to turn off Tailscale etc and normally this is fine. But this time, turning off Tailscale in the menu bar didn't actually solve the problem. `nslookup` on `captive.apple.com` gave `;; connection timed out; no servers could be reached` (same with `dig`). My iOS device with the same configuration did not have this problem.

<img width="600" alt="Image" src="https://github.com/user-attachments/assets/57c2bcd5-7ddd-45ff-92d5-fde54f08f940" />

I checked my iOS device and confirmed that DNS was automatically configured to `192.168.1.254`. I confirmed that I could ping this, but the ping results were kind of strange:

```
❯ ping 192.168.1.254
PING 192.168.1.254 (192.168.1.254): 56 data bytes
64 bytes from 192.168.1.254: icmp_seq=0 ttl=64 time=8.294 ms
64 bytes from 192.168.1.254: icmp_seq=0 ttl=64 time=8.344 ms (DUP!)
64 bytes from 192.168.1.254: icmp_seq=1 ttl=64 time=3.053 ms
64 bytes from 192.168.1.254: icmp_seq=1 ttl=64 time=3.177 ms (DUP!)
64 bytes from 192.168.1.254: icmp_seq=2 ttl=64 time=4.325 ms
64 bytes from 192.168.1.254: icmp_seq=2 ttl=64 time=4.808 ms (DUP!)
^C
--- 192.168.1.254 ping statistics ---
3 packets transmitted, 3 packets received, +3 duplicates, 0.0% packet loss
round-trip min/avg/max/stddev = 3.053/5.333/8.344/2.197 ms
```

Anyways, even when I could ping the DNS server, and after turning off VPN-on-demand, I still needed to restart wifi for the network to come back to a reasonable state. Then, I was able to reach `captive-portal.wifi-cloud.jp` via `captive.apple.com`.

Anyways, I ended up adding a Split DNS rule like what was mentioned here: https://github.com/tailscale/tailscale/issues/1634#issuecomment-2040239375

...and it worked the next time I got intercepted!

<img width="721" alt="Image" src="https://github.com/user-attachments/assets/34c26ef4-b1cb-4049-bebe-9c1b62b6e294" />

I still think there should be a more robust network check and detection scheme to try to detect situations like this. People like me who are well-versed in figuring out why stuff isn't working will probably eventually get network access. But it's very, very annoying and not seamless at all. The Split DNS settings might be a decent enough stop gap, but I don't see why it's necessary.
